src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -408,6 +408,23 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, BindingDi
             return result?.WithWasConverted();
         }
 
+#nullable enable
+        private BoundExpression BindToExtensionMemberOrInferredDelegateType(BoundMethodGroup methodGroup, BindingDiagnosticBag diagnostics)
+        {
+            // PROTOTYPE test use-site diagnostics
+            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+            MethodGroupResolution resolution = ResolveMethodGroup(methodGroup, analyzedArguments: null, isMethodGroupConversion: false, ref useSiteInfo);
+            diagnostics.Add(methodGroup.Syntax, useSiteInfo);
+            if (resolution.IsExtensionMember(out Symbol? extensionMember))
+            {
+                return GetExtensionMemberAccess(methodGroup.Syntax, methodGroup.ReceiverOpt, extensionMember, diagnostics);
+            }
+
+            // We have a method group so bind to an inferred delegate type
+            return BindToInferredDelegateType(methodGroup, diagnostics);
```

Comment:
>return BindToInferredDelegateType(methodGroup, diagnostics); [](http://example.com/codeflow?start=12&length=60)

Given that ```IsExtensionMember``` really means ```IsNonMethodExtensionMember```, is ```MethodGroup``` going to be mutated by ```ResolveMethodGroup``` when we are dealing with extension methods? It looks somewhat surprising that ```resolution``` isn't involved on this code path.

Reply:
> is MethodGroup going to be mutated by ResolveMethodGroup when we are dealing with extension methods?

I don't see how `ResolveMethodGroup` could mutate the `BoundMethodGroup` parameter. The members of `BoundMethodGroup` are readonly properties and the collections are readonly (`ImmutableArray`).

> It looks somewhat surprising that resolution isn't involved on this code path.

Yes, at the moment, we have separate logic for resolving method groups and for determining natural function type. 

FWIW, I have a related open design issue on this, where an instance method with bad arity could hide a good extension field. It's possible we could arrive at a way of unifying those two processes.